### PR TITLE
fix(auth): Store auth state data in redis

### PR DESCRIPTION
**Pull request by evanpurkhiser**
I considered separating out this logic into a seperate `StateStore` class, but it was simple enough to just implement in a few lines on the fetch and set state methods.

### CODE REVIEW

**Comment by getsentry-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Changes require @getsentry/security sign-off</td>
    </tr>
  </tbody>
</table>

### Security concerns found

- src/sentry/auth/helper.py


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

File: src/sentry/auth/helper.py [hunk 1]
```diff
@@ -41,6 +43,64 @@
 ERR_INVALID_IDENTITY = _('The provider did not return a valid user identity.')
 
 
+class RedisBackedState(object):
```
Comment (EvanPurkhiser) [inline src/sentry/auth/helper.py]: I'm not super happy with how much stuff is going on here, and feel like if we're going to have some type of generic redis session-cache type thing I might as well abstract it into a utility.. maybe I'll do that tomorrow.

File: src/sentry/auth/helper.py [hunk 2]
```diff
@@ -1,6 +1,7 @@
 from __future__ import absolute_import, print_function
 
 import logging
+import json
```
Comment (mattrobenolt) [inline src/sentry/auth/helper.py]: Use `sentry.utils.json` instead plz.

Comment (EvanPurkhiser) [reply src/sentry/auth/helper.py]: https://github.com/getsentry/sentry/pull/6245/commits/2e84910735a49fdb67c8598cbd7104507dd66e12

**Comment by EvanPurkhiser** [comment]
Will merge this tomorrow.

**Comment by EvanPurkhiser** [comment]
Could i get the final 🥇 on this when you have a chance @mattrobenolt. Wanted to get this merged in so we can finally avoid those opaque 502s.

**Review by mattrobenolt — APPROVED** [review]
:cookie:

**Pull request merged**